docs_src/design/adr/013-Device-Service-Events-Message-Bus.md

```diff
@@ -0,0 +1,257 @@
+  * [Configuration](#configuration)
+    + [Device Services](#device-services)
+      - [[MessageQueue]](#-messagequeue-)
+    + [Core Data](#core-data)
+      - [[MessageQueue]](#-messagequeue--1)
+    + [Application Services](#application-services-1)
+      - [[MessageBus]](#-messagebus-)
+      - [[Binding]](#-binding-)
+  * [Secure Connections](#secure-connections)
+- [Consequences](#consequences)
+
+## Status
+
+**proposed**
+
+## Context
+
+Currently EdgeX Events are sent from Device Services via HTTP to Core Data, which then puts the Events on the Message Bus. This ADR details how device services will send EdgeX Events to other services via the EdgeX Message Bus. 
+
+> *Note: Though this design is centered on device services, it does have cross cutting impacts with other EdgeX services and modules*
+
+> *Note: This ADR is depended on the [**Secret Provider for All**](TBD) to provide the secrets for secure Message Bus connections.*
+
+## Decision
+
+### Which Message Bus implementations?
+
+Multiple Device Services may need to be publishing Events to the Message Bus concurrently.  `ZMQ` will not be a valid option if multiple Device Services are configured to publish. This is because `ZMQ` only allows for a single publisher. `ZMQ` will still be valid if only one Device Service is publishing Events. The `MQTT` and `Redis Streams` are valid option to use with multiple Device Services publishing as they both support multiple publishers. These are the only other implementations currently available for Go services. The C base device services do not yet have a Message Bus implementation.  See the [C Device SDK](#c-device-sdk) below for details.
+
+> *Note: Documentation will need to be clear when `ZMQ` can be used and when it can not be used.*
+
+### Go Device SDK
+
+The Go Device SDK will take advantage of the existing `go-mod-messaging` module to enable use of the EdgeX Message Bus. A new bootstrap handler will be created which initializes the Message Bus client based on configuration. See [Configuration](#configuration) section below for details.  The Go Device SDK will be enhanced to optionally publish Events to the Message Bus anywhere it currently POSTs Events to Core Data. This publish vs POST option will be controlled by configuration with publish as the default.  See [Configuration](#configuration) section below for details. 
+
+### C Device SDK
+
+The C Device SDK will implement its own Message Bus abstraction similar to the one in `go-mod-messaging` with just the MQTT implementation to start with. Using this abstraction allows for future implementations to be added when use cases warrant the addition of `ZMQ` and/or `Redis Streams`.  As with the Go SDK, the C SDK will be enhanced to optionally publish Events to the Message Bus anywhere it currently POSTs Events to Core Data. This publish vs POST option will be controlled by configuration with publish as the default.  See [Configuration](#configuration) section below for details.
```

Comment:
This is up to Devices Service WG. @iain-anderson was focused first on `MQTT`, but I agree `Redis Streams` might be better first implementation. @iain-anderson Thoughts?

Reply:
I've done some work on MQTT already when that seemed to be the more useful option - to the point where it could be configured and messages were being sent. Getting to a working MQTT implementation probably just involves updating the message format.

Reply:
@iain-anderson , what are your thoughts on a `Redis Streams` implementation in the C SDK?

Reply:
Per TSC make this TBD decision.